sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala

```diff
@@ -60,23 +60,27 @@ class JdbcRelationProvider extends CreatableRelationProvider
 
     val conn = JdbcUtils.createConnectionFactory(jdbcOptions)()
     try {
-      val tableExists = JdbcUtils.tableExists(conn, url, table)
+      val tableSchema = JdbcUtils.getSchema(conn, url, table)
+      val tableExists = tableSchema.isDefined
+      val caseSensitive = sqlContext.conf.caseSensitiveAnalysis
       if (tableExists) {
         mode match {
           case SaveMode.Overwrite =>
-            if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
+            val savingSchema = if (isTruncate && isCascadingTruncateTable(url) == Some(false)) {
               // In this case, we should truncate table and then load.
               truncateTable(conn, table)
-              saveTable(df, url, table, jdbcOptions)
+              JdbcUtils.getSavingSchema(df.schema, tableSchema.get, caseSensitive)
             } else {
               // Otherwise, do not truncate the table, instead drop and recreate it
               dropTable(conn, table)
               createTable(df.schema, url, table, createTableOptions, conn)
-              saveTable(df, url, table, jdbcOptions)
+              df.schema
             }
+            saveTable(df, url, table, savingSchema, jdbcOptions)
```

Comment:
How about passing the table schema and resolve the schema inside `saveTable`? It might simplify the codes